allennlp/models/language_model.py

```diff
@@ -0,0 +1,206 @@
+from typing import Optional, Dict
+
+from overrides import overrides
+import torch
+import torch.nn as nn
+
+from allennlp.common import Params
+from allennlp.data import Vocabulary
+from allennlp.models.model import Model
+from allennlp.modules import TextFieldEmbedder
+from allennlp.common.checks import check_dimensions_match
+from allennlp.nn import InitializerApplicator, RegularizerApplicator
+
+from allennlp.modules.lm_rnn import LMRNN
+from allennlp.modules.adaptive import AdaptiveSoftmax
+from allennlp.training.metrics.perplexity import Perplexity
+
+@Model.register("WordLM")
+class WordLM(Model):
+    """
+    This ``WordLM`` is a word level language model.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``, required
+        A Vocabulary, required in order to compute sizes for input/output projections.
+    text_field_embedder : ``TextFieldEmbedder``, required
+        Used to embed the ``tokens`` ``TextField`` we get as input to the model.
+    rnn : ``LMRNN``, required
```

Comment:
Let's abstract this a little from RNNs to also allow other architectures such as transformer and CNN.  The only requirement on this module is that it is a `Seq2SeqEncoder`, so anything that satisfies that interface will work. An additional added benefit of this approach is that we already have wrappers for multilayer LSTMs in this interface, so most/all of the code in `lm_rnn.py` can be removed.

Reply:
I checked the current implementations of ```Seq2SeqEncoder```, and found the current wrapper would ***not*** wrap RNNs as stateful. However, RNNs based LM need RNNs to be stateful (and ```lm_rnn``` can be viewed as a simple version of stateful RNNs, which does not support mask). 

How about registering ```lm_rnn``` as a ```Seq2SeqEncoder```, and modify the class name from ```lm_rnn``` to ```simple_stateful_rnn```?

Reply:
We have a mechanism for stateful RNNs now -- https://github.com/allenai/allennlp/blob/master/allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py .   How does the functionality in `lm_rnn` differ from this?

Reply:
Yes, but the current ```__init.py``` does not turn on that option (https://github.com/allenai/allennlp/blob/master/allennlp/modules/seq2seq_encoders/__init__.py). Also, the mechanism for stateful RNNS requires the input having a ```mask``` field, which is not necessary for language modeling and would make the implementation slower.